File: Algo/Ruby/Recursion/get_permutations.rb
def getPermutations(array)
    return [[array[0]]] if array.length == 1
    collection = []
    sub_permutations = getPermutations(array[1, array.length - 1 ])
    sub_permutations.each do |sub_permutation|
        (0..sub_permutation.length).each do |index|
            copy = Array.new(sub_permutation)
            collection << copy.insert(index, array[0])
        end
    end

    return collection
end

array = [1,2,3,4]
p getPermutations(array)

File: Algo/JavaScript/Recursion/getPermutations.js
function getPermutations(array) {
    if (array.length === 0) {
        return [];
    }
    else if(array.length === 1) {
        return [[array[0]]]
    }
    let collection = [];
    let subPermutations = getPermutations(array.slice(1));
    for (const subPermutation of subPermutations) {
        for (let index = 0; index <= subPermutation.length; index++) {
            copy = Array.from(subPermutation);
            copy.splice(index, 0, array[0]);
            collection.push(copy);
        }
    }

    return collection;
}

array = [1,2,3,4,5,6,7];

console.log(getPermutations(array));
